File: SDK/python/verifio/bulk.py
from typing import List, Optional, cast, Dict, Any
from .client import VerifioClient
from .types import (
    BulkVerificationJob, 
    PaginationOptions, 
    VerificationResult, 
    PaginatedData
)

class BulkService:
    def __init__(self, client: VerifioClient):
        self.client = client

    def verify(self, emails: List[str]) -> BulkVerificationJob:
        body: Dict[str, Any] = {"emails": emails}
        
        response_data = self.client.request("POST", "/bulk", body)
        
        if not response_data.get("success") or not response_data.get("data"):
            raise Exception(response_data.get("error") or "Bulk verification failed")
            
        return response_data["data"]

    def get_job(self, job_id: str) -> BulkVerificationJob:
        response_data = self.client.request("GET", f"/bulk-jobs/{job_id}")
        
        if not response_data.get("success") or not response_data.get("data"):
            raise Exception(response_data.get("error") or "Failed to get job")
            
        return response_data["data"]

    def get_results(self, job_id: str, options: Optional[PaginationOptions] = None) -> PaginatedData[VerificationResult]:
        page = options.get("page", 1) if options else 1
        limit = options.get("limit", 20) if options else 20
        
        response_data = self.client.request("GET", f"/bulk-jobs/{job_id}/results?page={page}&limit={limit}")
        
        if not response_data.get("success") or not response_data.get("data"):
            raise Exception(response_data.get("error") or "Failed to get results")
            
        return response_data["data"]

File: SDK/rust/src/bulk.rs
use reqwest::Method;
use serde::Serialize;
use std::sync::Arc;

use crate::client::VerifioClient;
use crate::error::VerifioError;
use crate::types::{BulkVerificationJob, PaginatedData, PaginationOptions, VerificationResult};

#[derive(Clone, Debug)]
pub struct BulkService {
    client: Arc<VerifioClient>,
}

impl BulkService {
    pub(crate) fn new(client: Arc<VerifioClient>) -> Self {
        Self { client }
    }

    pub async fn verify(&self, emails: Vec<String>) -> Result<BulkVerificationJob, VerifioError> {
        #[derive(Serialize)]
        struct BulkRequest {
            emails: Vec<String>,
        }

        let body = BulkRequest { emails };

        self.client
            .request(Method::POST, "/bulk", Some(&body))
            .await
    }

    pub async fn get_job(&self, job_id: &str) -> Result<BulkVerificationJob, VerifioError> {
        let path = format!("/bulk-jobs/{}", job_id);
        self.client.request::<_, ()>(Method::GET, &path, None).await
    }

    pub async fn get_results(
        &self,
        job_id: &str,
        options: Option<PaginationOptions>,
    ) -> Result<PaginatedData<VerificationResult>, VerifioError> {
        let page = options.as_ref().and_then(|o| o.page).unwrap_or(1);
        let limit = options.as_ref().and_then(|o| o.limit).unwrap_or(20);

        let path = format!("/bulk-jobs/{}/results?page={}&limit={}", job_id, page, limit);

        self.client.request::<_, ()>(Method::GET, &path, None).await
    }
}
